src/main/java/net/minestom/server/particle/ParticleOptions.java

```diff
@@ -0,0 +1,148 @@
+package net.minestom.server.particle;
+
+import net.minestom.server.color.Color;
+import net.minestom.server.coordinate.Point;
+import net.minestom.server.entity.Entity;
+import net.minestom.server.item.ItemStack;
+import net.minestom.server.utils.MathUtils;
+import net.minestom.server.utils.binary.BinaryWriter;
+import net.minestom.server.utils.validate.Check;
+import org.jetbrains.annotations.NotNull;
+
+import static net.minestom.server.particle.ParticleOptions.*;
+
+public sealed abstract class ParticleOptions permits BlockStateOptions, Dust, DustColorTransition, Item, SculkCharge, Shriek, Vibration {
```

Comment:
This should be an interface, and the records should have an explicit type function (do not put the type function in the record constructor.)